### PR TITLE
Implementing line comments

### LINKED ISSUE #3112: Line comments like PromQL
**Is your feature request related to a problem? Please describe.**
Using the Loki datasource, the LogQL query editor in the _Explore_ view in Grafana syntax-highlights line comments beginning with (`#`); this is specified in the Loki datasource's [syntax description](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/loki/syntax.ts#L166).

image

However, as far as I can tell, this feature is not implemented in LogQL.

**Describe the solution you'd like**
Implement line comments in a way that matches the UI syntax highlighting and mirrors PromQL so as to not violate expectations.

**Describe alternatives you've considered**
Interestingly, line comments using `//` work and I would _guess_ we get this for free since `goyacc` is used? I can't see it explicitly mentioned in the [grammar definition](https://github.com/grafana/loki/blob/master/pkg/logql/expr.y).

**Additional context**
I have a branch where I've hacked together an implementation - if the maintainers would like this feature, I'll happily submit a PR.

### PR DESCRIPTION
**Pull request by dannykopping**
**What this PR does / why we need it**:
This PR implements line comments using the `#` character.

**Which issue(s) this PR fixes**:
Fixes #3112

**Special notes for your reviewer**:
I have not yet added documentation because it doesn't seem like we publish the LogQL grammar in the docs, and it feels a bit silly to document a line comment :upside_down_face: 

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Review by owen-d — COMMENTED** [review]
One little nit - I want to make sure we're not parsing valid filters that include `#` as comments now.

After that it looks good, but could you add a bit to the docs showing the new support?

File: pkg/logql/parser_test.go [hunk 1]
```diff
@@ -2128,6 +2128,47 @@ func TestParse(t *testing.T) {
...
+				left: newMatcherExpr([]*labels.Matcher{{Type: labels.MatchEqual, Name: "app", Value: "foo"}}),
+				pipeline: MultiStageExpr{
+					newLabelParserExpr(OpParserTypeJSON, ""),
+				},
+			},
+		},
+		{
+			in:  `#{app="foo"} | json`,
+			err: ParseError{msg: "syntax error: unexpected $end", line: 1, col: 20},
+		},
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply pkg/logql/parser_test.go:2171]: That's a good idea - I've added some further tests to this end

**Comment by dannykopping** [comment]
> One little nit - I want to make sure we're not parsing valid filters that include `#` as comments now.
> 
> After that it looks good, but could you add a bit to the docs showing the new support?

Done and done good sir :+1:

**Review by owen-d — APPROVED** [review]
LGTM, thanks

**Review by owen-d — APPROVED** [review]

**Pull request merged**